src/storage/src/render/upsert.rs

```diff
@@ -422,25 +429,49 @@ async fn drain_staged_input<S, G, T, O, E>(
     // This "mini-upsert" technique is actually useful in `UpsertState`'s
     // `merge_snapshot_chunk` implementation, minimizing gets and puts on
     // the `UpsertStateBackend` implementations. In some sense, its "upsert all the way down".
-    while let Some((ts, key, _, value)) = commands.next() {
-        let command_state = commands_state
-            .get_mut(&key)
-            .expect("key missing from commands_state");
+    while let Some((ts, key, from_time, value)) = commands.next() {
+        let mut command_state = if let Entry::Occupied(command_state) = commands_state.entry(key) {
+            command_state
+        } else {
+            panic!("key missing from commands_state");
+        };
+
+        let existing_value = &mut command_state.get_mut().value;
 
-        if let Some(cs) = command_state.value.as_mut() {
+        if let Some(cs) = existing_value.as_mut() {
             cs.ensure_decoded(bincode_opts);
         }
 
+        // In this commit, this is a check that will never occur, because of the `dedup_by` above.
+        let existing_order = existing_value.as_ref().map(|cs| cs.order().as_ref());
+
+        // If the value does not exist, or has the default `O` (None::<FromTime>`), we
+        // always sort below the from time of the new value. This ensures that new values
+        // overwrite keys with no values (and no tombstones), and all values that are
+        // recorded in rehydration (which also use the default `O`)
+        if existing_order >= Some(Some(&from_time.0)) {
+            command_state.remove();
+            continue;
+        }
```

Comment:
I can't parse this comment, where do we check for the default value? It also says "if X, we always sort".

Reply:
will adjust!